prog/hints.go

```diff
@@ -373,6 +380,44 @@ func shrinkExpand(v uint64, compMap CompMap, bitsize uint64, image bool) []uint6
 	return res
 }
 
+type HintsLimiter struct {
+	mu       sync.Mutex
+	attempts map[uint64]int // replacement attempts per PC
+}
+
+// Limit restricts hints to at most N replacement attempts per single kernel PC
+// (globally, across all hints mutations for all programs).
+// We are getting too many generated candidates, the fuzzer may not keep up
+// with them at all (hints jobs keep growing infinitely). If a hint indeed came
+// from the input w/o transformation, then we should guess it on the first
+// attempt (or at least after few attempts). If it did not come from the input,
+// or came with a non-trivial transformation, then any number of attempts won't
+// help. So limit the total number of attempts (until the next restart).
+func (limiter *HintsLimiter) Limit(comps CompMap) {
+	const N = 10
+	limiter.mu.Lock()
+	defer limiter.mu.Unlock()
+	if limiter.attempts == nil {
+		limiter.attempts = make(map[uint64]int)
+	}
+	for op1, ops2 := range comps {
+		for op2, pcs := range ops2 {
+			for pc := range pcs {
+				limiter.attempts[pc]++
```

Comment:
Did you count only the actual substitution attempts in the previous version of the code?

If yes, I think it may have become overly restrictive now. Far not every comparison from the map will lead to the actual substitution, so the 10 times limit has a much bigger effect here.

Reply:
No, it was the same code/logic before.